### PR TITLE
[Feature-5987][Server] Support to set multiple environment configs for a certain worker.

### LINKED ISSUE #5987: [Feature][Server] Support to set multiple environment configs for a certain worker.
**Describe the feature**

Now we can set an environment config in the file 'dolphinscheduler_env.sh' , but probably you need to modify this file in many servers. And a worker can't support to switch environment from one to another.So I expect that I can choose an environment config for different tasks.

**Describe the solution you'd like**
1.Add a web page to manage environments and create multiple environments in this page.
2.You can set some config like 'dolphinscheduler_env.sh' in a new environment and save it.
3.A certain worker group can have multiple environments, so need to link them together.
4.When set the config of a task, you may choose an environment to run this task.
5.Before running a task, the system will make up the file 'dolphinscheduler_env.sh' for this task.

### PR DESCRIPTION
**Pull request by calvinjiang**
## Purpose of the pull request

This PR closes #5987 .

Now we can set an environment config in the file 'dolphinscheduler_env.sh' , but probably you need to modify this file in many servers. And a worker can't support to switch environment from one to another. This PR provides new feature that you can choose an environment config for different tasks.

## Brief change log

1.Add a web page to manage environments and create multiple environments in this page.
2.You can set some config like 'dolphinscheduler_env.sh' in a new environment and save it.
image
3.An environment  can be linked with multiple worker groups, so need to link them together.
image
4.When set the config of a task, you may choose an environment to run this task.
image
5.When you manually run the process, you can choose one of environments that already linked with the worker group.
image
6.When you set a timer for your process, you can choose one of environments that already linked with the worker group.
image
5.Before running a task, the system will make up the file 'dolphinscheduler_env.sh' for this task.
image

## Verify this pull request

This change added tests and can be verified as follows:
  - *Manually verified the change by testing locally.*

### CODE REVIEW

**Comment by calvinjiang** [comment]
1、This  operator "===" can't be modified as "==", because the operator "==" breaks the rules of eslint.
image

2、These classes are all the JavaBean in the different levels. So they have many same properties and Get/Set methods.
image

File: dolphinscheduler-dao/pom.xml [hunk 1]
```diff
@@ -107,10 +108,11 @@
...
 
-        <dependency>
-            <groupId>mysql</groupId>
-            <artifactId>mysql-connector-java</artifactId>
-        </dependency>
+<!--        <dependency>-->
+<!--            <groupId>mysql</groupId>-->
+<!--            <artifactId>mysql-connector-java</artifactId>-->
+<!--            <scope></scope>-->
+<!--        </dependency>-->
```
Comment (zhuangchong) [inline dolphinscheduler-dao/pom.xml]: Please check the code here.

File: sql/dolphinscheduler_mysql.sql [hunk 2]
```diff
@@ -968,3 +975,36 @@ CREATE TABLE `t_ds_alert_plugin_instance` (
...
+  `code` int(11) DEFAULT NULL COMMENT 'encoding',
+  `name` varchar(100) NOT NULL COMMENT 'environment name',
+  `config` text NULL DEFAULT NULL COMMENT 'this config contains many environment variables config',
+  `description` text NULL DEFAULT NULL COMMENT 'the details',
+  `operator` int(11) DEFAULT NULL COMMENT 'operator user id',
+  `create_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
+  `update_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `environment_name_unique` (`name`),
+  UNIQUE KEY `environment_code_unique` (`name`),
```
Comment (zhuangchong) [inline sql/dolphinscheduler_mysql.sql]: There is a problem with the code here.

File: sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql [hunk 3]
```diff
@@ -632,6 +632,50 @@ d//
...
+   `code` int(11) DEFAULT NULL COMMENT 'encoding',
+   `name` varchar(100) NOT NULL COMMENT 'environment config name',
+   `config` text NULL DEFAULT NULL COMMENT 'this config contains many environment variables config',
+   `description` text NULL DEFAULT NULL COMMENT 'the details',
+   `operator` int(11) DEFAULT NULL COMMENT 'operator user id',
+   `create_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
+   `update_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+   PRIMARY KEY (`id`),
+   UNIQUE KEY `environment_name_unique` (`name`),
+   UNIQUE KEY `environment_code_unique` (`name`)
```
Comment (zhuangchong) [inline sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql]: same as above.

File: dolphinscheduler-dao/pom.xml (hunk 1 above)
Comment (calvinjiang) [reply dolphinscheduler-dao/pom.xml]: This problem has been fixed.

File: sql/dolphinscheduler_mysql.sql (hunk 2 above)
Comment (calvinjiang) [reply sql/dolphinscheduler_mysql.sql]: This problem has been fixed.

File: sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql (hunk 3 above)
Comment (calvinjiang) [reply sql/upgrade/1.4.0_schema/mysql/dolphinscheduler_ddl.sql]: This problem has been fixed.

**Review by zhuangchong — APPROVED** [review]
+1

**Review by kezhenxu94 — COMMENTED** [review]
Please also add the new database fields to https://github.com/apache/dolphinscheduler/blob/dev/sql/dolphinscheduler_h2.sql

**Comment by calvinjiang** [comment]
> Please also add the new database fields to https://github.com/apache/dolphinscheduler/blob/dev/sql/dolphinscheduler_h2.sql

These new tables and columns has added into the dolphinscheduler_h2.sql.

**Review by CalvinKirs — APPROVED** [review]
LGTM

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java [hunk 4]
```diff
@@ -185,4 +190,34 @@ public static String getPythonHome(String envPath) {
...
     }
 
+    /**
+     * get python home from the environment config
+     *
+     * @param environmentConfig env config
+     * @return python home
+     */
+    public static String getPythonHomeFromEnvironmentConfig(String environmentConfig) {
+        BufferedReader br = null;
```
Comment (ruanwenjun) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java]:
```suggestion
```
Remove the unused code.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java [hunk 5]
```diff
@@ -185,4 +190,34 @@ public static String getPythonHome(String envPath) {
...
+        StringBuilder sb = new StringBuilder();
+
+        String[] lines = environmentConfig.split("\n");
+
+        for (String line : lines) {
+            if (StringUtils.isNotEmpty(StringUtils.trim(line))) {
+                if (line.contains(Constants.PYTHON_HOME)) {
+                    sb.append(line);
```
Comment (ruanwenjun) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java]: It's better to use string here.
Or you can just use stream.filter() to find the first one.

File: sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql [hunk 6]
```diff
@@ -658,6 +658,64 @@ d//
...
+ALTER TABLE t_ds_task_instance ADD COLUMN environment_config text DEFAULT '';
+comment on column t_ds_task_instance.environment_config is 'environment config';
+
+--
+-- Table structure for table t_ds_environment_worker_group_relation
+--
+DROP TABLE IF EXISTS t_ds_environment_worker_group_relation;
+CREATE TABLE t_ds_environment_worker_group_relation (
+    id serial NOT NULL,
+    environment_code bigint NOT NULL,
```
Comment (ruanwenjun) [inline sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql:710]: It seems this field is int in mysql.

File: sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql [hunk 7]
```diff
@@ -658,6 +658,64 @@ d//
...
 SELECT uc_dolphin_T_t_ds_schedules_A_add_timezone();
 DROP FUNCTION IF EXISTS uc_dolphin_T_t_ds_schedules_A_add_timezone();
+
+--
+-- Table structure for table t_ds_environment
+--
+DROP TABLE IF EXISTS t_ds_environment;
+CREATE TABLE t_ds_environment (
+    id serial NOT NULL  ,
+    code bigint NOT NULL,
```
Comment (ruanwenjun) [inline sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql:668]: It seems this field is int in mysql.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java (hunk 4 above)
Comment (calvinjiang) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java]: Done

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java (hunk 5 above)
Comment (calvinjiang) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/PythonCommandExecutor.java]: done

File: sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql (hunk 6 above)
Comment (calvinjiang) [reply sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql:710]: done

File: sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql (hunk 7 above)
Comment (calvinjiang) [reply sql/upgrade/1.4.0_schema/postgresql/dolphinscheduler_ddl.sql:668]: done

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Pull request merged**